### PR TITLE
Enable range-based for loops for molecules

### PR DESCRIPTION
**Pull request by bp-kelley**
Adds range based for over atoms.

I think I can do the same for bonds as well if we like this approach.

### CODE REVIEW

File: Code/GraphMol/ROMol.h [hunk 1]
```diff
@@ -100,6 +100,35 @@ extern const int ci_ATOM_HOLDER;
...
+  struct CXXAtomIter {
+    Graph *graph;
+    typename Graph::vertex_iterator vend, pos;
+    Atom *current;
+    
+     CXXAtomIter(Graph *graph,
+                 typename Graph::vertex_iterator pos) : graph(graph), pos(pos), current(0) {}
+    
+    Vertex& operator*() { return current; }
+    CXXAtomIter& operator++() { ++pos; current = (*graph)[*pos].get(); return *this; }
```
Comment (bp-kelley) [inline Code/GraphMol/ROMol.h]: Something is fishy here, current needs to be assigned before ++

**Comment by bp-kelley** [comment]
So much nicer:

```
for (auto atom: atoms) {
  atom->getIdx();
}
```

**Comment by bp-kelley** [comment]
For the python fans:

```
#define in :

...

for(auto atom in atoms) {
  atom->getIdx();
}
```

**Comment by bp-kelley** [comment]
I'll need to valgrind this as well.

File: Code/GraphMol/ROMol.h [hunk 2]
```diff
@@ -175,6 +235,42 @@ class ROMol : public RDProps {
...
+         atom->getIdx();
+      };
+    \endcode
+   */
+  
+  CXXAtomIterator<MolGraph, Atom*> atoms() {
+        return { &d_graph };
+  }
+
+  CXXAtomIterator<const MolGraph, Atom*const> atoms() const {
```
Comment (greglandrum) [inline Code/GraphMol/ROMol.h]: I admit that I've never gotten my brain around "const pointer to const object" notation, but I think `Atom *const` should be `const Atom *const`.

File: Code/GraphMol/test1.cpp [hunk 3]
```diff
@@ -1357,10 +1357,61 @@ void testGithub1453() {
...
+
+    delete m;
+    
+}
+
 // -------------------------------------------------------------------
 int main() {
   RDLog::InitLogs();
 // boost::logging::enable_logs("rdApp.info");
+  /*
```
Comment (greglandrum) [inline Code/GraphMol/test1.cpp]: these tests need to be re-enabled.

File: Code/GraphMol/ROMol.h (hunk 2 above)
Comment (bp-kelley) [reply Code/GraphMol/ROMol.h]: I'll see if that compiles :)

Comment (greglandrum) [reply Code/GraphMol/ROMol.h]: Yeah, "ask the compiler" is generally my solution too.

**Comment by bp-kelley** [comment]
@coleb Care to take a look at the ranged based for implementation?  It looks straightforward, but there may be a better way.

@greglandrum @coleb Should this iterate through pointers or references?  I like pointers since there is no pass by value confusion, but I don't know the best practices for this.

File: Code/GraphMol/ROMol.cpp [hunk 4]
```diff
@@ -475,10 +475,10 @@ void ROMol::clearComputedProps(bool includeRings) const {
...
   RDProps::clearComputedProps();
 
-  for (ConstAtomIterator atomIt = this->beginAtoms();
-       atomIt != this->endAtoms(); ++atomIt) {
-    (*atomIt)->clearComputedProps();
+  for (auto atom: atoms()) {
+    atom->clearComputedProps();
   }
+  
   for (ConstBondIterator bondIt = this->beginBonds();
```
Comment (greglandrum) [inline Code/GraphMol/ROMol.cpp]: This could be `for (auto bond: bonds()){`, right?

**Comment by greglandrum** [comment]
Since the atoms/bonds are stored as pointers it seems like there's an argument for returning them, but references are, IMO, nicer if they work.
Still I think pointers isn't bad.

**Comment by coleb** [comment]
> I'm curious to see if there's a performance impact (positive or negative, but more likely the second?) of this. 

What I've heard is that range-based for loops can be optimized by the compiler better than the traditional syntax since the compiler is allowed to only call 'end()' once. If the compilers currently can't pull the 'endAtoms()' call outside the loop, and it currently has to be called on every loop iteration to check for termination, this could potentially be a big win. 

> @coleb Care to take a look at the ranged based for implementation? It looks straightforward, but there may be a better way.

Looks reasonable. May be able to hide the boost graph dependency a little big by using ROMol::operator[]?

> @greglandrum @coleb Should this iterate through pointers or references? I like pointers since there is no pass by value confusion, but I don't know the best practices for this.

I think pointers as we want to make the default performant. Accidentally creating copies seems bad by default: https://blog.knatten.org/2013/10/13/prefer-using-references-with-range-based-for-loops/

**Comment by greglandrum** [comment]
@coleb: sounds like you agree that it's fine to merge?

**Comment by coleb** [comment]
Yup, looks fine to me.

**Pull request merged**